### - Exercici 1
##### Agafa un text en anglès que vulguis, i calcula'n la freqüència de les paraules.

El texto que selecciono es un fragmento de una canción llamada 'Lover' de Taylor Swift. La frase dice 'Don't read the last page, but I stay when it's hard or it's wrong or we're making mistakes'. Refleja un compromiso emocional y estar presente en una relación a pesar de los desafios y los errores. 

In [2]:
import nltk
nltk.download('punkt')
nltk.download('stopwords')
from nltk.tokenize import word_tokenize
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\sampu\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\sampu\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [3]:
# Frase de la canción
frase = "Don't read the last page, but I stay when it's hard or it's wrong or we're making mistakes."

Ahora voy a preprocesar la frase, primero haré tokenización para dividir el texto en palabras. Después quitaré las mayúsculas y los caracteres especiales y puntuaciones. 

In [4]:
# Tokenización de palabras
tokens = nltk.word_tokenize(frase)
tokens

['Do',
 "n't",
 'read',
 'the',
 'last',
 'page',
 ',',
 'but',
 'I',
 'stay',
 'when',
 'it',
 "'s",
 'hard',
 'or',
 'it',
 "'s",
 'wrong',
 'or',
 'we',
 "'re",
 'making',
 'mistakes',
 '.']

In [5]:
# Converierto a minúsculas
tokens = [word.lower() for word in tokens]

In [6]:
# Eliminar puntuación y caracteres especiales
tokens = [word for word in tokens if word.isalnum()]
tokens

['do',
 'read',
 'the',
 'last',
 'page',
 'but',
 'i',
 'stay',
 'when',
 'it',
 'hard',
 'or',
 'it',
 'wrong',
 'or',
 'we',
 'making',
 'mistakes']

In [7]:
from collections import Counter

In [8]:
# Calcular la frecuencia de las palabras sin utilizar stopwords i stemming

word_freq = Counter(tokens)
print(word_freq)

Counter({'it': 2, 'or': 2, 'do': 1, 'read': 1, 'the': 1, 'last': 1, 'page': 1, 'but': 1, 'i': 1, 'stay': 1, 'when': 1, 'hard': 1, 'wrong': 1, 'we': 1, 'making': 1, 'mistakes': 1})


### - Exercici 2
##### Treu les stopwords i realitza stemming al teu conjunt de dades.

Para hacer el preprocesamiento completo voy a aplicar el ejercicio2, Stopwords, que son eliminar palabras como 'a', 'the', etc., que no aportan significado especial para el análisi, así se reduce el ruido de los datos. Y en el caso del Stemming, se reducen las palabras a su raíz o forma base. 

In [25]:
# Eliminar stopwords
stop_words = set(stopwords.words("english"))
tokens = [word for word in tokens if word not in stop_words]

In [26]:
# Stemming 
stemmer = PorterStemmer()
tokens = [stemmer.stem(word) for word in tokens]

In [27]:
tokens

['read', 'last', 'page', 'stay', 'hard', 'wrong', 'make', 'mistak']

Ahora después del preprocesamiento voy a calcular la frecuencia de las palabras. 

In [12]:
# Calcular la frecuencia de las palabras

word_freq = Counter(tokens)
print(word_freq)

Counter({'read': 1, 'last': 1, 'page': 1, 'stay': 1, 'hard': 1, 'wrong': 1, 'make': 1, 'mistak': 1})


### - Exercici 3
#### Realitza sentiment analysis al teu conjunt de dades.

Vamos a realizar el análisis de sentimiento para determinar la actitud emocional del texto. Ayuda a comprender y clasificar el tono emocional detrás del conjunto de palabras. 

In [18]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\sampu\AppData\Roaming\nltk_data...


True

In [22]:
# Inicializa el analizador de sentimientos de NLTK
analyzer = SentimentIntensityAnalyzer()

#calculamos el sentimiento de cada palabra
for word in tokens:
    print(word)
    print(analyzer.polarity_scores(word))
    print('\n')


read
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}


last
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}


page
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}


stay
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}


hard
{'neg': 1.0, 'neu': 0.0, 'pos': 0.0, 'compound': -0.1027}


wrong
{'neg': 1.0, 'neu': 0.0, 'pos': 0.0, 'compound': -0.4767}


make
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}


mistak
{'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}




In [24]:
#calculamos el sentimiento del texto
analyzer.polarity_scores(frase)


{'neg': 0.391, 'neu': 0.609, 'pos': 0.0, 'compound': -0.8402}

Finalmente nos da 4 sentimientos con un porcentaje, en función de la frase analizada. 

En este caso, lo que nos dice es:
- Existe un 39.1% se clasifica como negativo en términos de sentimiento.
- Un 60.9% que se clasifica como neutro. 
- Un 0% se clarifica como positivo. 

El compound es una métrica compuesta que resume la polaridad del texto en su conjunto. Este valor varía entre -1 (indicando muy negativo) y 1 (indicando muy positivo)

- Se entiende que el valor generadoees de -0.8402, por ello tiene una polariada negativa, ya que está más cercano a -1. 




CONCLUSIÓN: El análisis de sentimiento muestra que una parte pequeña del texto se clasifica como negativa, la mayoria como neutral y ninguna parte como positiva. En general el texto tiene una polaridad negativa. 